### PR TITLE
feat: support CSV output

### PR DESCRIPTION
**Pull request by sayon**
# What ❔

- Support for CSV output
- Refactoring to enable generation of JSON in LNT-friendly format

## Why ❔

- CSV output is required so that we can more easily analyze the performance reports using Excel, R, or similar tools.
- We are aiming at integrating with LNT.

## Checklist
 
- [X] PR title corresponds to the body of PR.
- [X] Tests for the changes have been added / updated.
- [X] Documentation comments have been added / updated.
- [X] Code has been formatted via `cargo fmt` and checked with `cargo clippy`.

### CODE REVIEW

File: compiler_tester/src/summary/benchmark_adapters/input.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+//!
+//! Converts `[InputIdentifier]` to the representation used by the benchmark.
+//!
+
+use crate::test::case::input::identifier::InputIdentifier;
+
+///
+/// Converts `[InputIdentifier]` to the representation used by the benchmark.
+///
+pub fn convert_input(input: InputIdentifier) -> benchmark_analyzer::Input {
```
Comment (hedgar2017) [inline compiler_tester/src/summary/benchmark_adapters/input.rs:10]: Could you change these into `From`/`Into`?

File: compiler_tester/src/test/context/input.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+//!
+
+use super::case::CaseContext;
+
+///
+/// Context used to process test inputs, organized in test cases.
+///
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct InputContext<'a> {
+    pub case_context: &'a CaseContext<'a>,
```
Comment (hedgar2017) [inline compiler_tester/src/test/context/input.rs:12]: Please add missing comments here.

File: benchmark_analyzer/src/benchmark/format/csv.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                        mode,
+                        group: _,
+                    },
+                size,
+                cycles,
+                ergs,
+                gas,
+            } in group.elements.values()
+            {
+                let size_str = size.map_or(String::from(""), |s| s.to_string());
```
Comment (hedgar2017) [inline benchmark_analyzer/src/benchmark/format/csv.rs:41]: Can be simplified to `size.map(|s| s.to_string()).unwrap_or_default()`?

File: compiler_tester/src/compiler_tester/arguments/benchmark_format.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+impl std::str::FromStr for BenchmarkFormat {
+    type Err = anyhow::Error;
+
+    fn from_str(string: &str) -> Result<Self, Self::Err> {
+        match string.to_lowercase().as_str() {
+            "json" => Ok(Self::Json),
+            "csv" => Ok(Self::Csv),
+            string => anyhow::bail!(
+                "Unknown benchmark format `{}`. Supported formats: {}",
```
Comment (hedgar2017) [inline compiler_tester/src/compiler_tester/arguments/benchmark_format.rs:17]: Please use `{string}` where possible!

File: benchmark_analyzer/src/benchmark/mod.rs [hunk 5]
```diff
@@ -202,10 +205,14 @@ impl Benchmark {
...
-    pub fn write_to_file(self, path: PathBuf) -> anyhow::Result<()> {
-        let contents = serde_json::to_string(&self).expect("Always valid");
+    pub fn write_to_file(
+        self,
+        path: PathBuf,
+        serializer: impl IBenchmarkSerializer,
+    ) -> anyhow::Result<()> {
+        let contents = serializer.serialize_to_string(&self).expect("Always valid");
         std::fs::write(path.as_path(), contents)
             .map_err(|error| anyhow::anyhow!("Benchmark file {:?} reading: {}", path, error))?;
```
Comment (hedgar2017) [inline benchmark_analyzer/src/benchmark/mod.rs:217]: Let's change it to `"Benchmark file {path:?} reading: {error}"`.

File: compiler_tester/src/compiler_tester/arguments/mod.rs [hunk 6]
```diff
@@ -40,6 +43,10 @@ pub struct Arguments {
     #[structopt(short, long)]
     pub benchmark: Option<PathBuf>,
 
+    /// The benchmark output format.
+    #[structopt(long = "benchmark-format")]
+    pub benchmark_format: Option<BenchmarkFormat>,
```
Comment (hedgar2017) [inline compiler_tester/src/compiler_tester/arguments/mod.rs:48]: Can it be changed to an `Option`-less arg with a default?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
It's an amazing piece of work, thank you!
I've left a few nits.

Have you tested all major use cases manually? I'd like to produce a benchmark of our LLVM EVM BE against *solc*. Is it able to compare reports such as `Y+M3B3` with `Y+`?

**Comment by sayon** [comment]
Fixed the comments! Let's merge when we can to minimize future conflicts.

LNT will be able to compare Y+M3B3 and Y+. But for now we can already do that using the CSV output, which is uniform regardless of the target.

For example, if I execute the following:
```
cargo run --bin compiler-tester -- --target evm --load-system-contracts system-contracts-stable-build --benchmark-format csv  -v  --benchmark  Y_evm.csv  --path tests/solidity/simple/default.sol 

cargo run --bin compiler-tester --  --target eravm  --load-system-contracts system-contracts-stable-build  -v  --benchmark-format csv --benchmark  Y.csv  --path tests/solidity/simple/default.sol
```

then I can import both files to a spreadsheet [for example to gdocs](https://docs.google.com/spreadsheets/d/19K6NI3Fpequd9X7EL7OisX-Wyvjm7CBx19loPBoL2U8/edit?gid=727563380#gid=727563380) and do basic statistics there. Once we plug in another backend to generate JSON for LNT, it can be automated.

File: compiler_tester/src/summary/benchmark_adapters/metadata.rs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+use crate::test::description::TestDescription;
+use crate::Mode;
+
+use super::selector::convert_selector;
+
+///
+/// Converts `[TestSelector]` to the representation used by the benchmark.
+///
+pub fn convert_description(
```
Comment (hedgar2017) [inline compiler_tester/src/summary/benchmark_adapters/metadata.rs:13]: impl From?

File: compiler_tester/src/compilers/solidity/upstream/mode.rs [hunk 8]
```diff
@@ -111,24 +111,29 @@ impl Mode {
...
 
-impl std::fmt::Display for Mode {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    ///
+    /// Returns a string representation excluding the solc version.
+    ///
+    pub fn repr_without_version(&self) -> String {
         if self.via_mlir {
-            return write!(f, "L {}", self.solc_version);
+            String::from("L")
```
Comment (hedgar2017) [inline compiler_tester/src/compilers/solidity/upstream/mode.rs:120]: Nit: still can do early return!

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

**Review by hedgar2017 — APPROVED** [review]

**Pull request merged**